torch_xla/experimental/scan.py

```
@@ -82,14 +88,335 @@ def scan(fn, init, xs):
+  flat_xs_len = len(flat_xs)
+
+  # `fn` operates on PyTrees and returns PyTrees. However, XLA only understands
+  # (lists or tuples of) tensors. So we will craft a `flat_fn` that takes in
+  # flattened PyTrees, internally recreates the desired tree structure, then calls `fn`.
+  def flat_fn(*seq: torch.Tensor) -> Tuple[PyTree, PyTree]:
+    carry = seq[:flat_carry_len]
+    x = seq[flat_carry_len:]
+    carry_pytree = tree_unflatten(carry, carry_spec)
+    x_pytree = tree_unflatten(x, xs_spec)
+    return fn(carry_pytree, x_pytree)
+
+  # Abstractly trace and lower `fn`.
+  # Later we will include `fn_computation` within the while loop body.
+  def make_fake_tensor(v: torch.Tensor) -> torch.Tensor:
+    # TODO(yifeit): there are some problems in PyTorch/XLA, or I'm missing something, because
+    #
+    #    torch.empty(v.size(), dtype=v.dtype, requires_grad=v.requires_grad).to(device)
+    #
+    # results in a tensor without gradient tracking, and
+    #
+    #    torch.empty(v.size(), dtype=v.dtype, device=v.device, requires_grad=v.requires_grad)
+    #
+    # results in incorrect calculation, unless they are `print()`-ed.
+    # But all three should be equivalent.
+    t = torch.empty(v.size(), dtype=v.dtype).to(device)
+    t.requires_grad_(v.requires_grad)
+    return t
+
+  device = torch_xla.device()
+  fake_carry_pytree = tree_map(make_fake_tensor, init)
+  fake_x_pytree = tree_map(lambda v: make_fake_tensor(v[0]), xs)
+  fake_carry, _ = tree_flatten(fake_carry_pytree)
+  fake_x, _ = tree_flatten(fake_x_pytree)
+  fn_output_carry_pytree, fn_output_y_pytree = flat_fn(*(fake_carry + fake_x))
+
+  # Later we'll use `fn_output_carry_spec` etc to turn flattened outputs back to a PyTree.
+  fn_output_carry, fn_output_carry_spec = tree_flatten(fn_output_carry_pytree)
+  assert fn_output_carry_spec == carry_spec
+  fn_output_y, fn_output_y_spec = tree_flatten(fn_output_y_pytree)
+  flat_y_len = len(fn_output_y)
+  fn_outputs = fn_output_carry + fn_output_y
```

what if there are in place updates to the tensor but it is not being return from the function?

---

I tested this and we give the wrong answer: https://github.com/tengyifei/playground/blob/master/scan_with_in_place_updates.ipynb

In the notebook, I wrote an approach to detect and prevent in place updates like that. TLDR is we'll have to trace every forward of each layer and verify that they're the same.

---

In the latest commit, this will now fail with an assertion error instead of silently giving wrong results.